tesseract-server/src/handlers/logic_layer/aggregate.rs

```diff
@@ -0,0 +1,222 @@
+use std::collections::HashMap;
+use std::convert::{TryInto};
+
+use actix_web::{
+    AsyncResponder,
+    FutureResponse,
+    HttpRequest,
+    HttpResponse,
+    Path,
+};
+use log::*;
+use serde_urlencoded;
+use futures::future::{self, Future};
+
+use tesseract_core::format::{format_records, FormatType};
+use tesseract_core::Query as TsQuery;
+
+use crate::app::AppState;
+use crate::handlers::logic_layer::shared::{LogicLayerQueryOpt, Time};
+
+
+/// Handles default aggregation when a format is not specified.
+/// Default format is CSV.
+pub fn logic_layer_default_handler(
+    (req, _cube): (HttpRequest<AppState>, Path<()>)
+) -> FutureResponse<HttpResponse>
+{
+    logic_layer_aggregation(req, "csv".to_owned())
+}
+
+/// Handles aggregation when a format is specified.
+pub fn logic_layer_handler(
+    (req, cube_format): (HttpRequest<AppState>, Path<(String)>)
+) -> FutureResponse<HttpResponse>
+{
+    logic_layer_aggregation(req, cube_format.to_owned())
+}
+
+
+/// Performs data aggregation.
+pub fn logic_layer_aggregation(
+    req: HttpRequest<AppState>,
+    format: String,
+) -> FutureResponse<HttpResponse>
+{
+    let format = format.parse::<FormatType>();
+    let format = match format {
+        Ok(f) => f,
+        Err(err) => {
+            return Box::new(
+                future::result(
+                    Ok(HttpResponse::NotFound().json(err.to_string()))
+                )
+            );
+        },
+    };
+
+    info!("format: {:?}", format);
+
+    let query = req.query_string();
+
+    let agg_query_res = serde_urlencoded::from_str::<Vec<(String, String)>>(query);
```

Comment:
I still strongly prefer the style:
```
struct LogicLayerQueryOpt {
  cube: String,
  dimensions: String,
  measures: String,
  #[serde(flatten)]
  cuts: HashMap<String, String>,
}
```
because
1) it provides a clearer view for maintainers to know what's in a query (I think data structures show this more clearly)
2) it provides one point of error handling, at deserialization, isntead of checking for each key

However, I leave this to you as I don't think there's anything wrong with using `HashMap` for all query params either.

Reply:
Oh, I think I also mentioned above, is there a reason you can't use `serde_qs`?